File: AgencyCampaign/AgencyCampaign.Domain/Entities/CampaignDocument.cs
using AgencyCampaign.Domain.ValueObjects;
using Archon.Core.Entities;

namespace AgencyCampaign.Domain.Entities
{
    public sealed class CampaignDocument : Entity
    {
        private readonly List<CampaignDocumentSignature> signatures = [];
        private readonly List<CampaignDocumentEvent> events = [];

        public long CampaignId { get; private set; }

        public Campaign? Campaign { get; private set; }

        public long? CampaignCreatorId { get; private set; }

        public CampaignCreator? CampaignCreator { get; private set; }

        public long? TemplateId { get; private set; }

        public CampaignDocumentTemplate? Template { get; private set; }

        public CampaignDocumentType DocumentType { get; private set; }

        public string Title { get; private set; } = string.Empty;

        public string? DocumentUrl { get; private set; }

        public string? Body { get; private set; }

        public string? Provider { get; private set; }

        public string? ProviderDocumentId { get; private set; }

        public string? SignedDocumentUrl { get; private set; }

        public CampaignDocumentStatus Status { get; private set; } = CampaignDocumentStatus.Draft;

        public string? RecipientEmail { get; private set; }

        public string? EmailSubject { get; private set; }

        public string? EmailBody { get; private set; }

        public DateTimeOffset? SentAt { get; private set; }

        public DateTimeOffset? SignedAt { get; private set; }

        public string? Notes { get; private set; }

        public IReadOnlyCollection<CampaignDocumentSignature> Signatures => signatures.AsReadOnly();

        public IReadOnlyCollection<CampaignDocumentEvent> Events => events.AsReadOnly();

        private CampaignDocument()
        {
        }

        public CampaignDocument(long campaignId, CampaignDocumentType documentType, string title, string? documentUrl = null, string? notes = null, long? campaignCreatorId = null, long? templateId = null, string? body = null)
        {
            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(campaignId);
            ArgumentException.ThrowIfNullOrWhiteSpace(title);

            CampaignId = campaignId;
            CampaignCreatorId = campaignCreatorId;
            TemplateId = templateId;
            DocumentType = documentType;
            Title = title.Trim();
            DocumentUrl = Normalize(documentUrl);
            Body = string.IsNullOrWhiteSpace(body) ? null : body;
            Notes = Normalize(notes);
        }

        public void Update(CampaignDocumentType documentType, string title, string? documentUrl, string? notes, long? campaignCreatorId)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(title);

            DocumentType = documentType;
            Title = title.Trim();
            DocumentUrl = Normalize(documentUrl);
            Notes = Normalize(notes);
            CampaignCreatorId = campaignCreatorId;
        }

        public void UpdateBody(string body)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(body);
            Body = body;
        }

        public void MarkReadyToSend()
        {
            Status = CampaignDocumentStatus.ReadyToSend;
        }

        public void MarkSent(string recipientEmail, string emailSubject, string? emailBody, DateTimeOffset sentAt)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(recipientEmail);
            ArgumentException.ThrowIfNullOrWhiteSpace(emailSubject);

            RecipientEmail = recipientEmail.Trim();
            EmailSubject = emailSubject.Trim();
            EmailBody = Normalize(emailBody);
            SentAt = sentAt.ToUniversalTime();
            Status = CampaignDocumentStatus.Sent;
        }

        public void AttachToProvider(string provider, string providerDocumentId)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(provider);
            ArgumentException.ThrowIfNullOrWhiteSpace(providerDocumentId);

            Provider = provider.Trim();
            ProviderDocumentId = providerDocumentId.Trim();

            if (Status == CampaignDocumentStatus.Draft || Status == CampaignDocumentStatus.ReadyToSend)
            {
                Status = CampaignDocumentStatus.Sent;
                SentAt ??= DateTimeOffset.UtcNow;
            }
        }

        public CampaignDocumentSignature AddSignature(CampaignDocumentSignerRole role, string signerName, string signerEmail, string? signerDocumentNumber = null, string? providerSignerId = null)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(signerName);
            ArgumentException.ThrowIfNullOrWhiteSpace(signerEmail);

            CampaignDocumentSignature signature = new(Id, role, signerName, signerEmail, signerDocumentNumber, providerSignerId);
            signatures.Add(signature);
            return signature;
        }

        public CampaignDocumentSignature? RegisterSignerSigned(string signerEmail, DateTimeOffset signedAt, string? ipAddress = null, string? userAgent = null, string? providerSignerId = null)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(signerEmail);

            CampaignDocumentSignature? signature = signatures.FirstOrDefault(item =>
                string.Equals(item.SignerEmail, signerEmail.Trim(), StringComparison.OrdinalIgnoreCase));

            if (signature is null && !string.IsNullOrWhiteSpace(providerSignerId))
            {
                signature = signatures.FirstOrDefault(item => item.ProviderSignerId == providerSignerId);
            }

            signature?.MarkSigned(signedAt, ipAddress, userAgent);
            return signature;
        }

        public void MarkViewed(DateTimeOffset viewedAt)
        {
            if (Status == CampaignDocumentStatus.Sent)
            {
                Status = CampaignDocumentStatus.Viewed;
            }
        }

        public void MarkSigned(DateTimeOffset signedAt, string? signedDocumentUrl = null)
        {
            SignedAt = signedAt.ToUniversalTime();
            Status = CampaignDocumentStatus.Signed;

            if (!string.IsNullOrWhiteSpace(signedDocumentUrl))
            {
                SignedDocumentUrl = signedDocumentUrl.Trim();
            }
        }

        public void MarkRejected(string? notes = null)
        {
            Notes = Normalize(notes);
            Status = CampaignDocumentStatus.Rejected;
        }

        public void MarkCancelled(string? notes = null)
        {
            Notes = Normalize(notes);
            Status = CampaignDocumentStatus.Cancelled;
        }

        public CampaignDocumentEvent RegisterEvent(CampaignDocumentEventType eventType, string? description = null, string? metadata = null, DateTimeOffset? occurredAt = null)
        {
            CampaignDocumentEvent evt = new(Id, eventType, description, metadata, occurredAt);
            events.Add(evt);
            return evt;
        }

        public bool AllSigned()
        {
            return signatures.Count > 0 && signatures.All(item => item.IsSigned);
        }

        private static string? Normalize(string? value)
        {
            return string.IsNullOrWhiteSpace(value) ? null : value.Trim();
        }
    }
}

File: AgencyCampaign/AgencyCampaign.Web/src/types/campaignDocument.ts
export const CampaignDocumentType = {
  CreatorAgreement: 1,
  BrandContract: 2,
  AuthorizationTerm: 3,
  BriefingAttachment: 4,
  Other: 5,
} as const
export type CampaignDocumentTypeValue =
  (typeof CampaignDocumentType)[keyof typeof CampaignDocumentType]

export const campaignDocumentTypeLabels: Record<CampaignDocumentTypeValue, string> = {
  1: 'Aceite do creator',
  2: 'Contrato da marca',
  3: 'Termo de autorização',
  4: 'Anexo de briefing',
  5: 'Outro',
}

export const CampaignDocumentStatus = {
  Draft: 1,
  ReadyToSend: 2,
  Sent: 3,
  Viewed: 4,
  Signed: 5,
  Rejected: 6,
  Cancelled: 7,
} as const
export type CampaignDocumentStatusValue =
  (typeof CampaignDocumentStatus)[keyof typeof CampaignDocumentStatus]

export const campaignDocumentStatusLabels: Record<CampaignDocumentStatusValue, string> = {
  1: 'Rascunho',
  2: 'Pronto para enviar',
  3: 'Enviado',
  4: 'Visualizado',
  5: 'Assinado',
  6: 'Recusado',
  7: 'Cancelado',
}

export const CampaignDocumentSignerRole = {
  Creator: 1,
  Agency: 2,
  Brand: 3,
  Other: 4,
} as const
export type CampaignDocumentSignerRoleValue =
  (typeof CampaignDocumentSignerRole)[keyof typeof CampaignDocumentSignerRole]

export const campaignDocumentSignerRoleLabels: Record<CampaignDocumentSignerRoleValue, string> = {
  1: 'Creator',
  2: 'Agência',
  3: 'Marca',
  4: 'Outro',
}

export const CampaignDocumentEventType = {
  Created: 1,
  ReadyToSend: 2,
  Sent: 3,
  Viewed: 4,
  SignerSigned: 5,
  Signed: 6,
  Rejected: 7,
  Cancelled: 8,
  ProviderSyncError: 9,
} as const
export type CampaignDocumentEventTypeValue =
  (typeof CampaignDocumentEventType)[keyof typeof CampaignDocumentEventType]

export const campaignDocumentEventTypeLabels: Record<CampaignDocumentEventTypeValue, string> = {
  1: 'Documento criado',
  2: 'Pronto para enviar',
  3: 'Enviado',
  4: 'Visualizado',
  5: 'Signatário assinou',
  6: 'Documento assinado',
  7: 'Recusado',
  8: 'Cancelado',
  9: 'Erro de sincronização',
}

export interface CampaignDocumentSignature {
  id: number
  role: CampaignDocumentSignerRoleValue
  signerName: string
  signerEmail: string
  signerDocumentNumber?: string
  providerSignerId?: string
  signedAt?: string
  ipAddress?: string
  isSigned: boolean
}

export interface CampaignDocumentEvent {
  id: number
  eventType: CampaignDocumentEventTypeValue
  occurredAt: string
  description?: string
  metadata?: string
}

export interface CampaignDocument {
  id: number
  campaignId: number
  campaignCreatorId?: number
  templateId?: number
  templateName?: string
  documentType: CampaignDocumentTypeValue
  title: string
  documentUrl?: string
  body?: string
  provider?: string
  providerDocumentId?: string
  signedDocumentUrl?: string
  status: CampaignDocumentStatusValue
  recipientEmail?: string
  emailSubject?: string
  emailBody?: string
  sentAt?: string
  signedAt?: string
  notes?: string
  createdAt: string
  updatedAt?: string
  signatures: CampaignDocumentSignature[]
  events: CampaignDocumentEvent[]
}
